app/features/contacts/contact-hooks.ts

```diff
@@ -0,0 +1,206 @@
+import type { RealtimePostgresChangesPayload } from '@supabase/supabase-js';
+import { useMutation, useQuery, useSuspenseQuery } from '@tanstack/react-query';
+import { useQueryClient } from '@tanstack/react-query';
+import type { QueryClient } from '@tanstack/react-query';
+import { useEffect, useMemo } from 'react';
+import { useLatest } from '~/lib/use-latest';
+import type { BoardwalkDbContact } from '../boardwalk-db/database';
+import { boardwalkDb } from '../boardwalk-db/database';
+import { useUserRef } from '../user/user-hooks';
+import type { Contact } from './contact';
+import { ContactRepository, useContactRepository } from './contact-repository';
+
+const contactsQueryKey = 'contacts';
+
+export class ContactsCache {
+  constructor(
+    private readonly queryClient: QueryClient,
+    private readonly userId: string,
+  ) {}
+
+  /**
+   * Adds a contact to the cache.
+   * @param contact - The contact to add.
+   */
+  add(contact: Contact) {
+    this.queryClient.setQueryData<Contact[]>(
+      [contactsQueryKey, this.userId],
+      (curr) => [...(curr ?? []), contact],
+    );
+  }
+
+  /**
+   * Gets all contacts in the cache for the current user.
+   * @returns The list of contacts.
+   */
+  getAll() {
+    return this.queryClient.getQueryData<Contact[]>([
+      contactsQueryKey,
+      this.userId,
+    ]);
+  }
+
+  /**
+   * Get a contact by id.
+   * @param id - The id of the contact.
+   * @returns The contact or null if the contact is not found.
+   */
+  get(id: string) {
+    const contacts = this.getAll();
+    return contacts?.find((x) => x.id === id) ?? null;
+  }
+
+  /**
+   * Removes a contact from the cache.
+   * @param contactId - The id of the contact to remove.
+   */
+  remove(contactId: string) {
+    this.queryClient.setQueryData<Contact[]>(
+      [contactsQueryKey, this.userId],
+      (curr) => curr?.filter((x) => x.id !== contactId) ?? [],
+    );
+  }
+}
+
+export function useContactsCache() {
+  const queryClient = useQueryClient();
+  const userRef = useUserRef();
+  return useMemo(
+    () => new ContactsCache(queryClient, userRef.current.id),
+    [queryClient],
+  );
+}
+
+type UseContactsOptions = {
+  filterFn?: (contact: Contact) => boolean;
+};
+
+/**
+ * Hook for listing contacts for the current user with optional filtering
+ */
+export function useContacts(options?: UseContactsOptions) {
+  const userRef = useUserRef();
+  const contactRepository = useContactRepository();
+  const contactsCache = useContactsCache();
+
+  const { filterFn } = options || {};
+
+  const { data: contacts } = useSuspenseQuery({
+    queryKey: [contactsQueryKey, userRef.current.id],
+    queryFn: async () => {
+      const allContacts = await contactRepository.getAll(userRef.current.id);
+
+      if (!filterFn) {
+        return allContacts;
+      }
+
+      return allContacts.filter(filterFn);
```

Comment:
I wonder if we should use useQuery select for data transformation instead of doing it in the queryFn. See this blog post that explains the diff https://tkdodo.eu/blog/react-query-data-transformations

in this current case it won't make much difference but might just be more powerful api that would later enable us to do more transformations